lib/services/local/access_list.go

```diff
@@ -161,73 +186,195 @@ func (a *AccessListService) GetAccessListsToReview(ctx context.Context) ([]*acce
+			if _, exists := originalOwnersMap[ownerName]; !exists {
+				if err := a.updateAccessListOwnerOf(ctx, accessList.GetName(), ownerName, true); err != nil {
+					return trace.Wrap(err)
+				}
+			}
+		}
+
+		// update references for old owners
+		for ownerName := range originalOwnersMap {
+			if _, exists := currentOwnersMap[ownerName]; !exists {
+				if err := a.updateAccessListOwnerOf(ctx, accessList.GetName(), ownerName, false); err != nil {
+					return trace.Wrap(err)
+				}
+			}
+		}
+
+		return nil
+	}
+
+	var actions []func() error
+
 	// If IGS is not enabled for this cluster we need to wrap the whole
 	// operation inside *another* lock so that we can accurately count the
 	// access lists in the cluster in order to prevent un-authorized use of
 	// the AccessList feature
-
-	action := updateAccessList
 	if !modules.GetModules().Features().GetEntitlement(entitlements.Identity).Enabled {
-		action = func() error {
-			err := a.service.RunWhileLocked(ctx, []string{createAccessListLimitLockName}, accessListLockTTL,
-				func(ctx context.Context, _ backend.Backend) error {
-					if err := a.VerifyAccessListCreateLimit(ctx, accessList.GetName()); err != nil {
-						return trace.Wrap(err)
-					}
-					return trace.Wrap(updateAccessList())
-				})
-			return trace.Wrap(err)
-		}
+		actions = append(actions, func() error { return a.VerifyAccessListCreateLimit(ctx, accessList.GetName()) })
 	}
 
-	if err := action(); err != nil {
-		return nil, trace.Wrap(err)
-	}
+	actions = append(actions, validateAccessList, updateAccessList, reconcileOwners)
+
+	err := a.service.RunWhileLocked(ctx, []string{accessListResourceLockName}, accessListLockTTL,
+		func(ctx context.Context, _ backend.Backend) error {
+			return a.service.RunWhileLocked(ctx, lockName(accessList.GetName()), accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+				for _, action := range actions {
+					if err := action(); err != nil {
+						return trace.Wrap(err)
+					}
+				}
+				return nil
+			})
+		})
 
-	return upserted, nil
+	return upserted, trace.Wrap(err)
 }
 
 // DeleteAccessList removes the specified access list resource.
 func (a *AccessListService) DeleteAccessList(ctx context.Context, name string) error {
-	err := a.service.RunWhileLocked(ctx, lockName(name), accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+	action := func(ctx context.Context, _ backend.Backend) error {
+		// Get list resource.
+		accessList, err := a.service.GetResource(ctx, name)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Check if the access list is a member or owner of any other access lists.
+		if len(accessList.Spec.MemberOf) > 0 {
+			return trace.AccessDenied("Cannot delete '%s', as it is a member of one or more other Access Lists", accessList.Spec.Title)
+		}
+		if len(accessList.Spec.OwnerOf) > 0 {
+			return trace.AccessDenied("Cannot delete '%s', as it is an owner of one or more other Access Lists", accessList.Spec.Title)
+		}
```

Comment:
good idea... I do still feel a bit unsure about those fields. I've tested every scenario I can think of and haven't seen them get out of whack, but I'm worried it *could* happen. is there a better way we could be keeping them updated, maybe listening for AccessListMemberCreated events or something instead?

Reply:
nevermind... found a few cases where they can get out of sync 😓 

fixing those now, I also think not erroring if they're updated is best, and just ignoring the passed-in values, and going with what's already saved (if updating), otherwise explicitly setting to empty arrays (if creating new list).